File: 238-product-of-array-except-self/238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n=nums.size();
        vector<int>pre(n,1);
        pre[0]=nums[0];
        vector<int>suff(n,1);
        suff[n-1]=nums[n-1];
        vector<int>ans(n,1);
        for(int i=1;i<n;i++){
            pre[i]=pre[i-1]*nums[i];
        }
        for(int i=n-2;i>=0;i--){
            suff[i]=suff[i+1]*nums[i];
        }
        ans[0]=suff[1];
        ans[n-1]=pre[n-2];
        for(int i=1;i<n-1;i++){
            ans[i]=pre[i-1]*suff[i+1];
        }
        return ans;
    }
};

File: 238-product-of-array-except-self/238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] arr){
        int[] right=new int[arr.length];
        int prod=1;
        
        for(int i=arr.length-1;i>=0;i--){
            prod *= arr[i];
            right[i]=prod;
        }
        
        prod=1;
        int[] res=new int[arr.length];
        for(int i=0;i<arr.length-1;i++){
            int lp=prod;
            int rp=right[i+1];
            
            res[i]=lp*rp;
            prod*=arr[i];
            
        }
        
        res[arr.length-1]=prod;
        return res;
    }
}